crates/core/src/transport/connection_handler.rs

```diff
@@ -0,0 +1,378 @@
+                                        .handle_unrecognized_remote(addr);
+                                }
+                            }
+                        }
+                        Err(e) => {
+                            tracing::warn!("Failed to receive UDP packet: {:?}", e);
+                        }
+                    }
+                },
+                // Handling of outbound packets
+                send_message = self.send_queue.recv() => {
+                    if let Some((remote_addr, event)) = send_message {
+                        match event {
+                            ConnectionEvent::SendRawPacket(data) => {
+                                if let Err(e) = self.socket.send_to(data.send_data(), remote_addr).await {
+                                    tracing::warn!("Failed to send UDP packet: {:?}", e);
+                                }
+                            }
+                            ConnectionEvent::ConnectionStart { remote_public_key, open_connection  }  => {
+                                match self.traverse_nat(remote_addr, remote_public_key).await {
+                                    Err(error) => {
+                                        tracing::error!(%error, ?remote_addr, "Failed to establish connection");
+                                    }
+                                    Ok(connection_info) => {
+                                        let (outbound_sender, outbound_receiver) = mpsc::channel(1);
+                                        let (inbound_sender, inbound_recv) = mpsc::channel(1);
+                                        self.connection_raw_packet_senders.insert(remote_addr, (connection_info, inbound_sender));
+                                        let _ = open_connection.send(Ok((outbound_sender, inbound_recv)));
+                                    }
+                                }
+                            }
+                        }
+                    }
+                },
+            }
+        }
+    }
+
+    async fn traverse_nat(
+        &mut self,
+        remote_addr: SocketAddr,
+        remote_public_key: TransportPublicKey,
+    ) -> Result<ConnectionInfo, TransportError> {
+        enum ConnectionState {
+            Start,
+            AckConnection,
+        }
+        // todo: probably should use exponential backoff with an upper limit: `timeout`
+        let timeout = Duration::from_secs(5);
+
+        // todo: probably instead of a fixed interval we should monotonically increase the interval
+        // until we reach a maximum, and then just keep trying at that maximum interval
+        let mut tick = tokio::time::interval(std::time::Duration::from_millis(200));
+        const MAX_FAILURES: usize = 20;
+        let mut failures = 0;
+        let mut packet = [0u8; MAX_PACKET_SIZE];
+        let mut state = ConnectionState::Start;
+
+        let inbound_sym_key_bytes = rand::random::<[u8; 16]>();
```

Comment:
Per `transport.md`, isn't the inbound_sym_key received from the other peer in an intro packet?

Reply:
No, it's generated locally and then sent, check the document; this inbound is the one used for decrypted messages sent to you, the one you receive is the otubound one to encrypt the messages sent to the other peer. 

You are probably confussed here because this is still not complete. In your Alice/Bob example, this code is the one executed by Alice at the beginning, Bob is missing here yet, that's the enxt thing I'll add.